lib/trino-filesystem/src/main/java/io/trino/filesystem/hdfs/HdfsFileSystem.java

```diff
@@ -125,4 +133,43 @@ public FileIO toFileIo()
     {
         return new ForwardingFileIo(this);
     }
+
+    @Override
+    public void batchDeleteFiles(Collection<String> paths)
+            throws IOException
+    {
+        if (!paths.isEmpty()) {
+            Map<String, List<String>> pathsGroupedByDirectory = paths.stream()
+                    .collect(groupingBy(path -> path.substring(0, path.lastIndexOf("/"))));
+            Set<String> directories = pathsGroupedByDirectory.keySet();
+            for (String directory : directories) {
+                Path directoryHadoopPath = hadoopPath(paths.stream().findFirst().get());
+                FileSystem fileSystem = environment.getFileSystem(context, directoryHadoopPath);
+                FileSystem rawFileSystem = getRawFileSystem(fileSystem);
+                if (rawFileSystem instanceof FileSystemWithBatchDelete fileSystemWithBatchDelete) {
+                    environment.doAs(context.getIdentity(), () -> {
+                        fileSystemWithBatchDelete.batchDeleteFiles(
+                                pathsGroupedByDirectory.get(directory).stream().map(HadoopPaths::hadoopPath).collect(toImmutableList()));
+                        return null;
+                    });
+                }
+                else {
+                    environment.doAs(context.getIdentity(), () -> {
+                        for (String path : pathsGroupedByDirectory.get(directory)) {
+                            fileSystem.delete(hadoopPath(path), false);
+                        }
+                        return null;
+                    });
+                }
+            }
+        }
+    }
+
+    public static FileSystem getRawFileSystem(FileSystem fileSystem)
+    {
+        if (fileSystem instanceof FilterFileSystem) {
```

Comment:
use enhanced instanceof

(i'm aware this code is copied)

Reply:
I should copied it smarter anyway :|